Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -358,6 +520,38 @@ bool TextureMapperLayer::shouldBlend() const
     return m_currentOpacity < 1;
 }
 
+bool TextureMapperLayer::flattensAsLeafOf3DSceneOr3DPerspective() const
+{
+    // Check if 3D scene leaf or if has perspective and has children and has 3D transformation
+    if (((!m_state.preserves3D && (m_parent && m_parent->preserves3D())) || m_layerTransforms.localTransform.hasPerspective())
+        && !m_children.isEmpty() && has3DLocalTransform())
+            return true;
+    return false;
+}
+
+bool TextureMapperLayer::has3DLocalTransform() const
+{
+    return !(!m_layerTransforms.localTransform.m13() && !m_layerTransforms.localTransform.m23()
+        && !m_layerTransforms.localTransform.m31() && !m_layerTransforms.localTransform.m32()
+        && m_layerTransforms.localTransform.m33() == 1 && !m_layerTransforms.localTransform.m34()
+        && !m_layerTransforms.localTransform.m43() && m_layerTransforms.localTransform.m44() == 1);
+}
+
+bool TextureMapperLayer::accumulatesTransformForChildren() const
+{
+    return preserves3D() || m_layerTransforms.localTransform.hasPerspective();
```

Comment:
Why do you need `hasPerspective` here. Do you have a test case?

Reply:
Not needed. Removed